pkg/reconciler/nodeconfig_reconciler.go

```diff
@@ -0,0 +1,413 @@
+			}
+		}
+	}
+
+	return nil
+}
+
+func countReferences(revision *v1alpha1.NetworkConfigRevision, configs []v1alpha1.NodeNetworkConfig) int {
+	refCnt := 0
+	for j := range configs {
+		if configs[j].Spec.Revision == revision.Spec.Revision {
+			refCnt++
+		}
+	}
+	return refCnt
+}
+
+func (ncr *NodeConfigReconciler) listConfigs(ctx context.Context) (*v1alpha1.NodeNetworkConfigList, error) {
+	nodeConfigs := &v1alpha1.NodeNetworkConfigList{}
+	if err := ncr.client.List(ctx, nodeConfigs); err != nil {
+		return nil, fmt.Errorf("error listing nodeConfigs: %w", err)
+	}
+	return nodeConfigs, nil
+}
+
+func (ncr *NodeConfigReconciler) deployNodeConfig(ctx context.Context, node *corev1.Node, revision *v1alpha1.NetworkConfigRevision) error {
+	newConfig, err := ncr.createConfigForNode(node, revision)
+	if err != nil {
+		return fmt.Errorf("error preparing config for node %s: %w", node.Name, err)
+	}
+	currentConfig := &v1alpha1.NodeNetworkConfig{}
+	if err := ncr.client.Get(ctx, types.NamespacedName{Name: node.Name}, currentConfig); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("error getting NodeNetworkConfig object for node %s: %w", node.Name, err)
+		}
+		currentConfig = nil
+	}
+	if currentConfig != nil && currentConfig.Spec.Revision == revision.Spec.Revision {
+		// current config is the same as current revision - skip
+		return nil
+	}
+	if err := ncr.deployConfig(ctx, newConfig, currentConfig, node); err != nil {
+		if errors.Is(err, InvalidConfigError) || errors.Is(err, context.DeadlineExceeded) {
+			// revision results in invalid config or in context timeout - invalidate revision
+			revision.Status.IsInvalid = true
+			if err := ncr.client.Status().Update(ctx, revision); err != nil {
+				return fmt.Errorf("error invalidating revision %s: %w", revision.Name, err)
+			}
+		}
+		return fmt.Errorf("error deploying config for node %s: %w", node.Name, err)
+	}
+	return nil
+}
+
+func (ncr *NodeConfigReconciler) createConfigForNode(node *corev1.Node, revision *v1alpha1.NetworkConfigRevision) (*v1alpha1.NodeNetworkConfig, error) {
+	// create new config
+	c := &v1alpha1.NodeNetworkConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: node.Name,
+		},
+	}
+
+	c.Spec = *revision.Spec.Config.DeepCopy()
+	c.Spec.Revision = revision.Spec.Revision
+	c.Name = node.Name
+
+	if err := controllerutil.SetOwnerReference(node, c, scheme.Scheme); err != nil {
+		return nil, fmt.Errorf("error setting owner references (node): %w", err)
+	}
+
+	if err := controllerutil.SetOwnerReference(revision, c, ncr.scheme); err != nil {
+		return nil, fmt.Errorf("error setting owner references (revision): %w", err)
+	}
+
+	// prepare Layer2NetworkConfigurationSpec (l2Spec) for each node.
+	// Each Layer2NetworkConfigurationSpec from l2Spec has node selector,
+	// which should be used to add config to proper nodes.
+	// Each Layer2NetworkConfigurationSpec that don't match the node selector
+	// is removed.
+	for i := 0; i < len(c.Spec.Layer2); i++ {
```

Comment:
you could use [`slices.DeleteFunc`](https://pkg.go.dev/slices#DeleteFunc) instead, probably a bit more readable.

Reply:
Done